website/addons/onedrive/views.py

```diff
@@ -0,0 +1,183 @@
+@must_have_addon('onedrive', 'user')
+@must_have_addon('onedrive', 'node')
+@must_be_addon_authorizer('onedrive')
+@must_have_permission(permissions.WRITE)
+def onedrive_set_config(node_addon, user_addon, auth, **kwargs):
+    """View for changing a node's linked onedrive folder."""
+    folder = request.json.get('selected')
+    serializer = OneDriveSerializer(node_settings=node_addon)
+
+    logger.debug('folder::' + repr(folder))
+    logger.debug('serializer::' + repr(serializer))
+
+    name = folder['name']
+
+    node_addon.set_folder(folder, auth=auth)
+
+    return {
+        'result': {
+            'folder': {
+                'name': name,
+                'path': name,
+            },
+            'urls': serializer.addon_serialized_urls,
+        },
+        'message': 'Successfully updated settings.',
+    }
+
+
+@must_have_addon('onedrive', 'user')
+@must_have_addon('onedrive', 'node')
+@must_have_permission(permissions.WRITE)
+def onedrive_add_user_auth(auth, node_addon, user_addon, **kwargs):
+    """Import onedrive credentials from the currently logged-in user to a node.
+    """
+    external_account = ExternalAccount.load(
+        request.json['external_account_id']
+    )
+
+    if external_account not in user_addon.external_accounts:
+        raise HTTPError(http.FORBIDDEN)
+
+    try:
+        node_addon.set_auth(external_account, user_addon.owner)
+    except PermissionsError:
+        raise HTTPError(http.FORBIDDEN)
+
+    node_addon.set_user_auth(user_addon)
+    node_addon.save()
+
+    return {
+        'result': OneDriveSerializer().serialize_settings(node_addon, auth.user),
+        'message': 'Successfully imported access token from profile.',
+    }
+
+
+@must_not_be_registration
+@must_have_addon('onedrive', 'node')
+@must_have_permission(permissions.WRITE)
+def onedrive_remove_user_auth(auth, node_addon, **kwargs):
+    node_addon.deauthorize(auth=auth)
+    node_addon.save()
+
+
+@must_have_addon('onedrive', 'user')
+@must_have_addon('onedrive', 'node')
+@must_have_permission(permissions.WRITE)
+def onedrive_get_share_emails(auth, user_addon, node_addon, **kwargs):
+    """Return a list of emails of the contributors on a project.
+
+    The current user MUST be the user who authenticated OneDrive for the node.
+    """
+    if not node_addon.user_settings:
+        raise HTTPError(http.BAD_REQUEST)
+    # Current user must be the user who authorized the addon
+    if node_addon.user_settings.owner != auth.user:
+        raise HTTPError(http.FORBIDDEN)
+
+    return {
+        'result': {
+            'emails': [
+                contrib.username
+                for contrib in node_addon.owner.contributors
+                if contrib != auth.user
+            ],
+        }
+    }
+
+
+@must_have_addon('onedrive', 'node')
+@must_be_addon_authorizer('onedrive')
+def onedrive_folder_list(node_addon, **kwargs):
+    """Returns a list of folders in OneDrive"""
+    if not node_addon.has_auth:
```

Comment:
You'd have no way of knowing this, but we're planning on doing a small refactor to how folder-type addons list their folders soon -- we want this to be something that can happen outside of a flask context.

Please move the bulk of this to a function on the `OneDriveNodeSettings` class, and have this view 
`return node_addon.folder_list(<folder_id extracted from flask's request.args>)`